lib_eio/core/dla.ml

```diff
@@ -0,0 +1,14 @@
+let prepare_for_await () =
+  let state = Atomic.make `Init in
+  let release () =
+    if Atomic.get state != `Released then
+      match Atomic.exchange state `Released with
+      | `Awaiting enqueue -> enqueue (Ok ())
+      | _ -> ()
+  and await () =
+    if Atomic.get state != `Released then
+      Suspend.enter @@ fun _ctx enqueue ->
+      if not (Atomic.compare_and_set state `Init (`Awaiting enqueue)) then
+        enqueue (Ok ())
```

Comment:
The cancellation support seems to have gone missing. I would expect something like this:
```suggestion
      Suspend.enter @@ fun ctx enqueue ->
      let awaiting = `Awaiting enqueue in
      if Atomic.compare_and_set state `Init awaiting then (
        Cancel.Fiber_context.set_cancel_fn ctx (fun ex ->
            if Atomic.compare_and_set state awaiting `Released then (
              enqueue (Error ex)
            )
          )
      ) else (
        enqueue (Ok ())
      )
```

Reply:
Hmm... Is the `cancel_fn` always cleared automatically in case the fiber is enqueued normally?

Reply:
LGTM